hal/src/nRF52840/ble_hal.cpp

```diff
@@ -2487,6 +2555,27 @@ int BleObject::GattClient::init() {
         CHECK(gattcDiscPool_.init(GATT_CLIENT_DISC_SVC_CHAR_POOL_SIZE));
         gattcDiscPoolInit_ = true;
     }
+    if (os_semaphore_create(&discoverySemaphore_, 1, 0)) {
+        discoverySemaphore_ = nullptr;
+        LOG(ERROR, "os_semaphore_create() failed");
+        return SYSTEM_ERROR_INTERNAL;
+    }
+    if (os_semaphore_create(&readSemaphore_, 1, 0)) {
+        readSemaphore_ = nullptr;
+        os_semaphore_destroy(discoverySemaphore_);
+        discoverySemaphore_ = nullptr;
+        LOG(ERROR, "os_semaphore_create() failed");
+        return SYSTEM_ERROR_INTERNAL;
+    }
+    if (os_semaphore_create(&writeSemaphore_, 1, 0)) {
+        writeSemaphore_ = nullptr;
+        os_semaphore_destroy(discoverySemaphore_);
```

Comment:
Let's use a dismissable `NAME_SCOPE_GUARD` here or at least a goto label that will destroy these semaphores once in case of an error.